zcash_client_backend/src/data_api/chain.rs

```diff
@@ -224,6 +225,195 @@ pub trait BlockSource {
+/// #        2u32.into(),
+/// #        zcash_primitives::block::BlockHash([0; 32]),
+/// #        sapling::Nullifier([0; 32]),
+/// #        &dfvk,
+/// #        zcash_primitives::transaction::components::amount::NonNegativeAmount::const_from_u64(5),
+/// #        false,
+/// #        None,
+/// #    );
+///    let compact_blocks = vec![compact_block1, compact_block2];
+///
+///    // Insert blocks into the block cache
+///    block_cache.insert(compact_blocks.clone()).unwrap();
+///    assert_eq!(block_cache.cached_blocks.lock().unwrap().len(), 2);
+///
+///    // Find highest block in the block cache
+///    let get_tip_height = block_cache.get_tip_height(None).unwrap();
+///    assert_eq!(get_tip_height, Some(BlockHeight::from_u32(2)));
+///
+///    // Read from the block cache
+///    let blocks_from_cache = block_cache.read(&range).unwrap();
+///    assert_eq!(blocks_from_cache, compact_blocks);
+///
+///    // Truncate the block cache
+///    block_cache.truncate(BlockHeight::from_u32(1)).unwrap();
+///    assert_eq!(block_cache.cached_blocks.lock().unwrap().len(), 1);
+///    assert_eq!(
+///        block_cache.get_tip_height(None).unwrap(),
+///        Some(BlockHeight::from_u32(1))
+///    );
+///
+///    // Delete blocks from the block cache
+///    let rt = tokio::runtime::Runtime::new().unwrap();
+///    rt.block_on(async {
+///        block_cache.delete(&range).await.unwrap();
+///    });
+///    assert_eq!(block_cache.cached_blocks.lock().unwrap().len(), 0);
+///    assert_eq!(block_cache.get_tip_height(None).unwrap(), None);
+/// ```
+#[async_trait]
+pub trait BlockCache: BlockSource + Send + Sync {
+    /// Retrieves contiguous compact blocks specified by the given `range` from the block cache.
+    ///
+    /// Returns `Ok(Vec<CompactBlock>)` on success, otherwise returns an error.
+    ///
+    /// Short reads are allowed, meaning that returning fewer blocks than requested should not
+    /// return an error as long as all blocks are sequentially continuous in height.
+    ///
+    /// # Errors
+    ///
+    /// This method should return an error if there are gaps in the requested range of blocks,
+    /// indicating there are blocks missing from the cache.
+    fn read(&self, range: &ScanRange) -> Result<Vec<CompactBlock>, Self::Error>;
+
+    /// Finds the height of the highest block known to the block cache within a specified range.
+    /// If `range` is `None`, returns the tip of the entire cache.
+    ///
+    /// Returns `Ok(Some(BlockHeight))` on success, otherwise returns an error.
+    /// If no blocks are found in the cache, returns Ok(`None`).
+    fn get_tip_height(&self, range: Option<&ScanRange>)
+        -> Result<Option<BlockHeight>, Self::Error>;
+
+    /// Inserts a vec of compact blocks into the block cache.
+    ///
+    /// Returns `Ok(())` on success, otherwise returns an error.
+    fn insert(&self, compact_blocks: Vec<CompactBlock>) -> Result<(), Self::Error>;
```

Comment:
Should the mutating methods take `&mut`? That seems morally correct, even if the inner cache uses `Rc` or `Arc` to manage the database, but is there an issue preventing this that I'm not aware of?

Also, looking at this trait again, this method and `read` need to be async. The implementation of this trait cannot just modify the block metadata database; it also needs to do the disk / memory writes. So any mutations need to be async in the same way that deletions are, to support whatever the cache needs to do. (The previous join handle approach meant that I wasn't thinking about the overall async process, and was just thinking about the "deletion deferrment", but the operations themselves can also be async.)

Reply:
I agree with making more functions async. I don't recall needing async for reads in your implementation but for inserts I did change for a sync version when I was trying to find a solution for not using async traits. Either way, making all these async will support a wider range of implemetations and sync optimisations. That being said, making insert async will also require some test framework such as `generate_blocks_at` to also become async or be refactored, although this is not a problem to implement at all.

Regarding the `&mut` I think it would be good to have quick call soon to discuss the sync implementation especially when it comes to concurrency. For example, maybe im missing something, but would there not be issues with running several `delete` tasks that take a `&mut` as in your 1169-sync-engine? Or would there be issues for example reading from the cache for scanning while concurrently downloading and inserting into the cache which would be a good sync optimisation?